Blog post: Quinoa - A Quarkus extension to create Modern UI with no hassles (#1436)
quarkusio/quarkusio.github.io: _posts/2022-05-19-quinoa-modern-ui-with-no-hassles.adoc

---
layout: post

> title: 'Quinoa - A Quarkus extension to create Modern UI with no hassles'

Reviewer: ```suggestion
title: 'Quinoa - A Quarkus extension to create Modern UI with no hassle'
```
and you need to update the file name too

> title: 'Quinoa - A Quarkus extension to create Modern UI with no hassles'

Reviewer 2: ```suggestion
title: 'Quinoa - A Quarkiverse extension to create a web UI with no hassles'
```

date: 2022-05-19

> tags: quinoa web ui react angular vue lit spa nodejs frontend extension

Reviewer 2: please don't add that many tags and try use existing ones.

too many unique ones makes the right side tag menu overcrowded.

Author: Maybe the problem is with the menu :) but ok...

Reviewer 2: sure but until then "overloading" tags is problematic.

> synopsis: Quinoa extension takes care of all the Web UI build/wiring/dev-mode hassles to let you focus of your Quarkus web application logic.

Reviewer: ```suggestion
synopsis: The Quinoa extension takes care of all the Web UI build/wiring/dev-mode hassles to let you focus of your Quarkus web application logic.
```

author: adamevin
---
:imagesdir: /assets/images/posts/quinoa

image::quinoa.jpg[alt=Quinoa Bowl]

Quinoa... To be honest I wasn't really a fan at first, but then my wife made them in a bowl like a salad with avocado and raw salmon. Now I really enjoy it with a touch of soy sauce. Let's not digress too much, I am not here to talk about healthy food in this post, this is about an extension to enjoy coding modern UI with Quarkus.

== Get started

Create a new Quinoa project (with a base Quinoa starter code):

* With https://code.quarkus.io/?a=quinoa-bowl&j=17&e=io.quarkiverse.quinoa%3Aquarkus-quinoa[code.quarkus.io]
* With the https://quarkus.io/guides/cli-tooling[Quarkus CLI]:
[source,bash]
----
quarkus create app quinoa-app -x=io.quarkiverse.quinoa:quarkus-quinoa
----

Then start the live-coding *and navigate to http://0.0.0.0:8080/quinoa.html*:
[source,bash]
----
quarkus dev
----

https://quarkiverse.github.io/quarkiverse-docs/quarkus-quinoa/dev/#getting-started[Here] is the Quinoa guide to go deeper and configure your app with any Node based Web UI framework such as React, Angular, Lit, Webpack, Rollup, ...

=== A bit of context

> Alongside Quarkus, interacting with web frameworks such as React and Angular, is a natural need. Many users have been requesting a smooth integration since the early stages of Quarkus (https://github.com/quarkusio/quarkus/issues/588[#588]). This is why I created the Quinoa extension for Quarkus.

Reviewer: ```suggestion
Alongside Quarkus, interacting with web frameworks such as React and Angular is a natural need. Many users have been requesting a smooth integration since the early stages of Quarkus (https://github.com/quarkusio/quarkus/issues/588[#588]). This is why I created the Quinoa extension for Quarkus.
```

> Since then (3 years), link:#existing-articles[many blog articles] have popped up to help in that matter, most of them taking advantage of the frontend-maven-plugin. As creator of code.quarkus.io, I have also been confronted with this situation and I, like many others, have filled the gap using this plugin. It does the job for sure and has been good enough to fill the need for the time being. Unfortunately the overall process could be more streamlined and easier to get started with, it involves a lot of configuration, some tricky setup to copy the resources to the right place and test the frontend at the right moment. Live coding requires starting two processes and a proxy to make it work together.

Reviewer: `Since then` is weird? `Since the creation of Quarkus`? Or `publication`?

> Since then (3 years), link:#existing-articles[many blog articles] have popped up to help in that matter, most of them taking advantage of the frontend-maven-plugin. As creator of code.quarkus.io, I have also been confronted with this situation and I, like many others, have filled the gap using this plugin. It does the job for sure and has been good enough to fill the need for the time being. Unfortunately the overall process could be more streamlined and easier to get started with, it involves a lot of configuration, some tricky setup to copy the resources to the right place and test the frontend at the right moment. Live coding requires starting two processes and a proxy to make it work together.

Reviewer: ```suggestion
Since then (3 years), link:#existing-articles[many blog articles] have popped up to help in that matter, most of them taking advantage of the frontend-maven-plugin. As the creator of code.quarkus.io, I have also been confronted with this situation and I, like many others, have filled the gap using this plugin. It does the job for sure and has been good enough to fill the need for the time being. Unfortunately the overall process could be more streamlined and easier to get started with, it involves a lot of configuration, some tricky setup to copy the resources to the right place and test the frontend at the right moment. Live coding requires starting two processes and a proxy to make it work together.
```

=== Let's start cooking

> With this experience behind me, I was in a good position to contribute something to ease our lives a bit. I put an eye to all the needs, issues, propositions and blog articles and came up with a new extension to create healthy modern user-interfaces (UI) with Quarkus: Quinoa (standing for Quarkus UI with NO hAssles :)

Reviewer: ```suggestion
With this experience behind me, I was in a good position to contribute something to ease our lives a bit. I put an eye to all the needs, issues, propositions and blog articles and came up with a new extension to create healthy modern user-interfaces (UI) with Quarkus: Quinoa (standing for Quarkus UI with NO hAssle :).
```

> Quinoa is taking care of all the Web UI build/wiring/live-config hassles to let you take care of your web application logic. Either to create a full stack Single Page Application or Microservice (micro-frontend); things are made easier for you.

Reviewer: ```suggestion
Quinoa is taking care of all the Web UI build/wiring/live-config tasks to let you take care of your web application logic. Either to create a full stack Single Page Application or Microservice (micro-frontend); things are made easier for you.
```

> Quinoa is taking care of all the Web UI build/wiring/live-config hassles to let you take care of your web application logic. Either to create a full stack Single Page Application or Microservice (micro-frontend); things are made easier for you.

Reviewer 2: the sentence before says there is no hassles now quninoa is doing hassles :)

```suggestion
Quinoa is taking care of hiding all the hassels around Web UI build/wiring/live-config to let you take care of your web application logic. Whether you create a full-stack Single Page Application or Microservice (micro-frontend); things are made easier for you.
```

> NodeJS has great tooling to create modern frontend web applications. Whilst Quarkus has a broad ecosystem to build reliable modern backends in the cloud (microservices, rest, DBs, events, cloud native…). Quinoa is the glue to get the best of both worlds.

Reviewer: ```suggestion
Node.js has great tooling to create modern frontend web applications. Whilst Quarkus has a broad ecosystem to build reliable modern backends in the cloud (microservices, rest, DBs, events, cloud native…). Quinoa is the glue to get the best of both worlds.
```
(do a search and replace, it's a registered trademark)

> NodeJS has great tooling to create modern frontend web applications. Whilst Quarkus has a broad ecosystem to build reliable modern backends in the cloud (microservices, rest, DBs, events, cloud native…). Quinoa is the glue to get the best of both worlds.

Reviewer 2: ```suggestion
NodeJS has great tooling to create modern frontend web applications. Quarkus has a broad ecosystem to build reliable modern backends in the cloud (microservices, rest, DBs, events, cloud native…). Quinoa is the glue to get the best of both worlds.
```

> The only prerequisite for the NodeJS project to be compatible with Quinoa is to contain the `build` script to build the Web UI and deliver static files in a specific directory. Since teams can be different, to make everyone happy, it’s possible to either locate the UI in the Quarkus project (default is `src/main/webui`) or outside (as soon as it’s available on the disk at build time).

Author: Fixed

It is not necessary (but possible) to https://quarkiverse.github.io/quarkiverse-docs/quarkus-quinoa/dev/#package-manager[configure the package manager] (NPM, Yarn or PNPM), it will be auto-detected depending on the project lockfile.

> Then Quinoa will take care of the generated UI static files, they will be bundled in the jar or native binary and efficiently served alongside your backend (by Vert.x). Also something that UI developers were asking was to enable Single Page App routing. It is now just a https://quarkiverse.github.io/quarkiverse-docs/quarkus-quinoa/dev/#spa-routing[boolean configuration].

Reviewer 2: ```suggestion
Quinoa will take care of the generated UI static files. They will be bundled in the jar or native binary and efficiently served alongside your backend (by Vert.x). UI developers also were asking was to enable Single Page App routing. It is now just a https://quarkiverse.github.io/quarkiverse-docs/quarkus-quinoa/dev/#spa-routing[boolean configuration].
```

does the "efficiently" matter here btw. anything specially done elsewhere where it is "ineffeciently" served?

Now, what about live coding? It’s a popular feature of Quarkus and most NodeJS UI frameworks already have such a mode.
Quinoa provides two options:

* Delegate to the UI live coding server. https://quarkiverse.github.io/quarkiverse-docs/quarkus-quinoa/dev/#dev-server[You configure the port of the UI server]. By convention Quinoa will call the `start` script from the package.json to start the UI server process. Then it will transparently proxy relevant requests to the given port.
* Quarkus watches the files and Quinoa triggers a new Web UI build on changes (you can configure different builds for dev and prod).

> Either way, when starting Quarkus dev-mode your Web UI will be served seamlessly (by default on http://0.0.0.0:8080). Both are viable solutions and allow you to live-code on both backend and frontend at the same time with close to no setup! Pretty awesome right?

Reviewer: ```suggestion
Either way, when starting Quarkus dev-mode your Web UI will be served seamlessly (by default on http://0.0.0.0:8080). Both are viable options and allow you to live-code on both backend and frontend at the same time with close to no setup! Pretty awesome right?
```
